### PR TITLE
feat: optimize editor performance of selection service and scroll service.

### PR DESCRIPTION
**Pull request by LucasXu0**
## Rendering
* Replace the layout from `Column` with `ListView`. Only render the visible node displayed on the current screen.

## Selection service
* Only search selection from the visible node.
* Migrate the function of rendering the selection area and cursor to each selectable node.

## scroll service,
* Using the built-in scrollable from `ListView`.
* Using the `EdgeDraggingAutoScroller` to optimize the scroll event.

### CODE REVIEW

**Comment by LucasXu0** [comment]
The PR is for optimizing the performance of rendering a large document. A rough comparison video shows below.

The first half is the recording of release/0.1.1, while the second half is the recording of this branch. It can be seen that whether it is scrolling, computing selection area, or editing, the performance is much better than the previous version.

https://user-images.githubusercontent.com/11863087/227879574-c3154086-2eeb-4c55-b505-7f46b3ac02ff.mp4

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/blocks/base_component/rich_text_with_selection.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import 'package:flutter/material.dart';
+
+class RichTextWithSelection extends StatefulWidget {
+  const RichTextWithSelection({
+    super.key,
+    required this.text,
+    this.textSelection,
+    this.selectionColor = const Color.fromARGB(100, 33, 149, 243),
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/blocks/base_component/rich_text_with_selection.dart:8]: Very arbitrary, maybe we should consider using EditorTheme or something at some point.

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/editor/editor_entry.dart [hunk 2]
```diff
@@ -30,28 +27,106 @@ class EditorNodeWidget extends StatelessWidget {
...
+                  editorState: widget.editorState,
+                ),
+        );
+      },
+      itemCount: children.length,
+    );
     return Column(
       crossAxisAlignment: CrossAxisAlignment.start,
-      children: node.children
+      children: widget.node.children
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/editor/editor_entry.dart:70]:
```suggestion
      children: children
```

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart [hunk 3]
```diff
@@ -80,4 +88,95 @@ class _RichTextNodeWidgetState extends State<RichTextNodeWidget>
...
+      ..clear();
+    if (selection == null) {
+      return;
+    }
+    selection = selection.normalized;
+    final path = widget.textNode.path;
+    if (path < selection.start.path || path > selection.end.path) {
+      return;
+    }
+    var newSelection = selection.copyWith();
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart:117]:
```suggestion
    var newSelection = selection;
```

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart [hunk 4]
```diff
@@ -80,4 +88,95 @@ class _RichTextNodeWidgetState extends State<RichTextNodeWidget>
...
+              rect: rect,
+              color: Colors.blue.withOpacity(0.3),
+            );
+          },
+        );
+        _selectionOverlays.add(entry);
+      }
+      Overlay.of(context)?.insertAll(_selectionOverlays);
+    }
+  }
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart:175]:
```suggestion
    if (!selection.isSingle) {
      if (path.equals(newSelection.start.path)) {
        newSelection = newSelection.copyWith(end: end());
      } else if (path.equals(newSelection.end.path)) {
        newSelection = newSelection.copyWith(start: start());
      } else {
        newSelection = Selection(start: start(), end: end());
      }
    }

    final rects = getRectsInSelection(newSelection);
    if (rect == null) {
      return;
    }

    for (final rect in rects) {
      final entry = OverlayEntry(
        builder: (context) {
          return SelectionWidget(
            layerLink: widget.textNode.layerLink,
            rect: rect,
            color: Colors.blue.withOpacity(0.3),
          );
        },
      );
      _selectionOverlays.add(entry);
    }

    Overlay.of(context)?.insertAll(_selectionOverlays); 
  }
```
Maybe something like this?

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/selection/selectable.dart [hunk 5]
```diff
@@ -66,4 +67,6 @@ mixin SelectableMixin<T extends StatefulWidget> on State<T> {
   bool get shouldCursorBlink => true;
 
   CursorStyle get cursorStyle => CursorStyle.verticalLine;
+
+  void setSelection(Selection? selection) {}
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/selection/selectable.dart:71]:
```suggestion
  void setSelection(Selection? selection) => UnimplementedError('setSelection has not been implemented');
```

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/service/selection_service.dart [hunk 6]
```diff
@@ -272,8 +289,14 @@ class _AppFlowySelectionState extends State<AppFlowySelection>
...
-    final sortedNodes =
-        editorState.document.root.children.toList(growable: false);
+    // final sortedNodes =
+    //     editorState.document.root.children.toList(growable: false);
+    final sortedNodes = [
+      ...selectables.toList(growable: false)
+        ..sort((a, b) {
+          return a.path <= b.path ? -1 : 1;
+        })
+    ];
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/appflowy_editor/lib/src/service/selection_service.dart:299]:
```suggestion
    final sortedNodes = selectables.toList(growable: false)
        ..sort((a, b) {
          return a.path <= b.path ? -1 : 1;
        });
```

**Comment by Xazin** [comment]
I know it's draft, but I looked it over anyway, cheers :+1:

File: frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart (hunk 4 above)
Comment (LucasXu0) [reply frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart:175]: Hey, @Xazin. This PR is a draft. Just to validate the logic. I will format and refactor all the code once this logic is validated without any issues.

**Pull request closed without merging**